ledger/complete/mtrie/flattener/encoding.go

```diff
@@ -1,187 +1,456 @@
+func ReadNode(reader io.Reader, scratch []byte, getNode func(nodeIndex uint64) (*node.Node, error)) (*node.Node, error) {
+
+	// minBufSize should be large enough for interim node and leaf node with small payload.
+	// minBufSize is a failsafe and is only used when len(scratch) is much smaller
+	// than expected.  len(scratch) is 4096 by default, so minBufSize isn't likely to be used.
+	const minBufSize = 1024
+
+	if len(scratch) < minBufSize {
+		scratch = make([]byte, minBufSize)
 	}
 
-	version, _, err := utils.ReadUint16(buf)
+	// fixed-length data: node type (1 byte) + height (2 bytes) + max depth (2 bytes) + reg count (8 bytes)
+	const fixLengthSize = 1 + 2 + 2 + 8
+
+	// Read fixed-length part
+	pos := 0
+
+	_, err := io.ReadFull(reader, scratch[:fixLengthSize])
 	if err != nil {
-		return nil, fmt.Errorf("error reading storable node: %w", err)
+		return nil, fmt.Errorf("failed to read fixed-length part of serialized node: %w", err)
 	}
 
-	if version > encodingDecodingVersion {
-		return nil, fmt.Errorf("error reading storable node: unsuported version %d > %d", version, encodingDecodingVersion)
+	// Decode node type (1 byte)
+	nType := scratch[pos]
+	pos++
+
+	if nType != byte(leafNodeType) && nType != byte(interimNodeType) {
+		return nil, fmt.Errorf("failed to decode node type %d", nType)
 	}
 
-	// reading fixed-length part
-	buf = make([]byte, 2+8+8+2+8)
+	// Decode height (2 bytes)
+	height := binary.BigEndian.Uint16(scratch[pos:])
+	pos += 2
 
-	read, err = io.ReadFull(reader, buf)
-	if err != nil {
-		return nil, fmt.Errorf("error reading storable node, cannot read fixed-length part: %w", err)
-	}
-	if read != len(buf) {
-		return nil, fmt.Errorf("not enough bytes read %d expected %d", read, len(buf))
+	// Decode max depth (2 bytes)
+	maxDepth := binary.BigEndian.Uint16(scratch[pos:])
+	pos += 2
+
+	// Decode reg count (8 bytes)
+	regCount := binary.BigEndian.Uint64(scratch[pos:])
+
+	if nType == byte(leafNodeType) {
+
+		// Read encoded hash data and create hash.Hash.
+		nodeHash, err := readHashFromReader(reader, scratch)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read and decode hash of serialized node: %w", err)
+		}
+
+		// Read encoded path data and create ledger.Path.
+		path, err := readPathFromReader(reader, scratch)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read and decode path of serialized node: %w", err)
+		}
+
+		// Read encoded payload data and create ledger.Payload.
+		payload, err := readPayloadFromReader(reader, scratch)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read and decode payload of serialized node: %w", err)
+		}
+
+		node := node.NewNode(int(height), nil, nil, path, payload, nodeHash, maxDepth, regCount)
+		return node, nil
 	}
 
-	storableNode := &StorableNode{}
+	// Read interim node
+
+	pos = 0
 
-	storableNode.Height, buf, err = utils.ReadUint16(buf)
+	// Read left and right child index (8 bytes each)
+	_, err = io.ReadFull(reader, scratch[:16])
```

Comment:
isn't it we already have read 13 bytes of the scratch up to this point ?

Reply:
hmm, but don't we need to start reading a full value at index 13 instead of 0
?

Reply:
> don't we need to start reading a full value at index 13 instead of 0?

No because we can overwrite the scratch buffer content after decoding its previous contents.  For example,

1. we read 13 bytes from reader to `scratch[0:13]`
2. we decode data in `scratch[0:13]` (such as node type height, reg count, max depth), so at this point we don't need anything else from `scratch[0:13]`
3. we read 16 bytes from reader to `scratch[0:16]`
4. we decode two uint64 from `scratch[0:16]`